board: arm: Add Support For 96Boards Aerocore2

ric96 opened the pull request:
Changes:
- Added all required board files in /boards/arm/96b_aerocore2
- Modified pinmux for stm32f4
- Add stm32f427 support based on previous work done for the stm32f429.
- Rework currunt stm32f429 implimentation to now be based on stm32f427.
- Introduce dedicated dtsi for the VI variant of both stm32f427 and stm32f429. This is done to prevent stm32f4.dtsi from being included twice.

Most of the changes in this PR is based on reverse-engineering the PCB
layout and following commits in the PX4 firmware repository for the
same board. The manufacturer does not provide and or generate schematics
and pinout tables for this board.

This PR includes almost all of the interfaces connected to the STM32
MCU, the only thing not included is the J9 and J8 headers that connect
to a 96Boards baseboard. These headers are not vital to the functionality
of the Aerocore2.

Signed-off-by: Sahaj Sarup <sahaj.sarup@linaro.org>

ric96 commented:
not sure what the rtc error is about :/

erwango commented on boards/arm/96b_aerocore2/96b_aerocore2.yaml:
```diff
@@ -0,0 +1,21 @@
+  - gpio
+  - ble
+  - i2c
+  - counter
```
If specifying support for counter, you need to enable rtc node in board dts

erwango commented on dts/arm/st/f4/stm32f427vX.dtsi:
```diff
@@ -0,0 +1,11 @@
+
+/delete-node/ &spi5;
+
+/delete-node/ &spi6;
```
I'm not sure why this would be needed

ric96 replied:
f427's almost all variants have spi5 and spi6 but not the vX. I followed suit from the f429 implementation.

erwango replied:
Ok.

erwango reviewed (changes requested):
Thanks for this addition, few comments.
Also, can you please split into 2 commits (soc and board)?

ric96 commented on dts/arm/st/f4/stm32f427.dtsi:
```diff
@@ -0,0 +1,7 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <st/f4/stm32f429.dtsi>
```
Okay, yes so on this note. The thought did come to mind to either base 27 on the already present 29 or redo 27 as if from scratch and then base 29 on it?

Also in the case what happens to the Licences headers?

erwango replied:
>  redo 27 as if from scratch and then base 29 on it

Yes

> Also in the case what happens to the License headers?

They'll be both Linaro, so that's just a matter of year. You can move both to 2020!

ric96 commented on boards/arm/96b_aerocore2/96b_aerocore2.dts:
```diff
@@ -0,0 +1,111 @@
+/dts-v1/;
+
+#include <st/f4/stm32f427Xi.dtsi>
+#include <st/f4/stm32f427vX.dtsi>
```
well. it is a f427-vi part and the only modification for vX parts is the deletion of spi 5 & 6

erwango replied:
If we follow the inclusion paths, you're including twice stm32f4.dtsi

ric96 replied:
I don't know how one would go solving this. There are vX variants that are not Xi and vice versa. This is the same thing being done for the f429vi.

The only I can think of is to create a separate dtsi for vi parts, but that just seems redundant.

erwango replied:
For sure there will be redundancy somewhere, so let's find the minimal one.
I propose: ( with ">" to be read as "is included by" )
stm32f427 > stm32f427vX > stm32f427vi
                  > stm32f427Xi
                  > stm32f429 > stm32f429vX > stm32f429 vi
                  > stm32f429 > stm32f429Xi

ric96 committed:
arm: soc: add support for stm32f427

Add STM32F427. This is mainly aimed towards the stm32f427vi.

Changes:

- Add stm32f427 support based on previous work
done for the stm32f429.
- Rework currunt stm32f429 implimentation to now
be based on stm32f427.
- Introduce dedicated dtsi for the VI variant of both
stm32f427 and stm32f429. This is done to prevent stm32f4.dtsi
from being included twice.

Signed-off-by: Sahaj Sarup <sahaj.sarup@linaro.org>

ric96 committed:
board: arm: Add Support For 96Boards Aerocore2

Changes:

- Added all required board files in /boards/arm/96b_aerocore2
- Modified pinmux for stm32f4

Most of the changes in this PR is based on reverse-engineering of the
PCB layout and following commits in the PX4 firmware repository for
the same board. The manufacturer does not provide and or generate
schematics and pinout tables for this board.

This PR includes almost all of the interfaces connected to the STM32
MCU, the only thing not included is the J9 and J8 headers that connect
to a 96Boards baseboard.
These headers are not vital to the functionality of the Aerocore2.

Signed-off-by: Sahaj Sarup <sahaj.sarup@linaro.org>

erwango commented on boards/arm/96b_aerocore2/96b_aerocore2.dts:
```diff
@@ -0,0 +1,110 @@
+	leds {
+		compatible = "gpio-leds";
+		yellow_led_1: led_1 {
+			gpios = <&gpioe 10 GPIO_INT_ACTIVE_HIGH>;
```
Please rebase on latest maser and update vs new GPIO API.

ric96 replied:
done.

erwango commented on boards/arm/96b_aerocore2/Kconfig.defconfig:
```diff
@@ -0,0 +1,69 @@
+	default y
+
+endif # I2C
+
```
Add ADC, cf https://github.com/zephyrproject-rtos/zephyr/pull/22375

erwango commented on boards/arm/96b_aerocore2/96b_aerocore2.yaml:
```diff
@@ -0,0 +1,20 @@
+  - pwm
+  - i2c
+  - spi
+  - usb_device
```
adc

erwango commented on boards/arm/96b_aerocore2/doc/index.rst:
```diff
@@ -0,0 +1,345 @@
++------------+------------+-------------------------------------+
+| PWM        | on-chip    | timers / pwm                        |
++------------+------------+-------------------------------------+
+| USB OTG FS | on-chip    | USB device                          |
```
Please add ADC

erwango commented on boards/arm/96b_aerocore2/96b_aerocore2_defconfig:
```diff
@@ -0,0 +1,46 @@
+# Copyright (c) 2020 Linaro Limited
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_ARM=y
```
ARM is now deduced from SoC selection.
Setting it in board _defconfig generates an error (this is what is screwing our shippable results)

erwango commented on boards/arm/96b_aerocore2/96b_aerocore2_defconfig:
```diff
@@ -0,0 +1,46 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_ARM=y
+CONFIG_BOARD_96B_AEROCORE2=y
```
CONFIG_BOARD_XXX should now be removed from board _defconfig

erwango commented on boards/arm/96b_aerocore2/pinmux.c:
```diff
@@ -0,0 +1,107 @@
+	{STM32_PIN_PC1, STM32F4_PINMUX_FUNC_PC1_ADC123_IN11},
+	{STM32_PIN_PC2, STM32F4_PINMUX_FUNC_PC2_ADC123_IN12},
+	{STM32_PIN_PC3, STM32F4_PINMUX_FUNC_PC3_ADC123_IN13},
+#endif	/* CONFIG_ADC_1 */
```
USB is enabled in dts, but USB pins assignment is missing here

erwango commented on boards/arm/96b_aerocore2/Kconfig.defconfig:
```diff
@@ -0,0 +1,76 @@
+
+if UART_CONSOLE
+
+config UART_7
```
Please apply recently changed https://github.com/zephyrproject-rtos/zephyr/pull/22666

erwango commented on boards/arm/96b_aerocore2/Kconfig.defconfig:
```diff
@@ -0,0 +1,65 @@
+
+config UART_2
+	default y
+        depends	on SERIAL
```
A tab poped up here (and below)?

erwango reviewed (changes requested):
Minor fix

erwango commented:
^^ @galak,@MaureenHelm

MaureenHelm commented on dts/arm/st/f4/stm32f429.dtsi:
```diff
@@ -1,90 +1,7 @@
 /*
- * Copyright (c) 2017 Linaro Limited
+ * Copyright (c) 2020 Linaro Limited
```
Don't remove an existing copyright. Change to:
```
Copyright (c) 2017, 2020 Linaro Limited
```

ric96 replied:
fixed

MaureenHelm commented on boards/arm/96b_aerocore2/Kconfig.defconfig:
```diff
@@ -0,0 +1,65 @@
+
+config UART_7
+	default y
+	depends on SERIAL
```
Wrap in an `if SERIAL..endif` instead of repeating `depends on SERIAL`. Do this for PWM and SPI too

ric96 replied:
I was requested to make this change following https://github.com/zephyrproject-rtos/zephyr/pull/22666
https://github.com/zephyrproject-rtos/zephyr/pull/22095#discussion_r378376970

MaureenHelm replied:
#22666 was for single symbol if's. Here you have 4 UART instance symbols that all depend on the same thing.

MaureenHelm commented on boards/arm/96b_aerocore2/doc/index.rst:
```diff
@@ -0,0 +1,347 @@
+********
+
+The 96Boards Aerocore2 Mezzanine is based on the STMicroelectronics STM32F427VIT6 Cortex-M4 CPU primarily designed for use in drones.
+\This board acts as a mezzanine platform for all 96Boards CE compliant boards. It can also be used as a standalone board.
```
Please fix the long lines

ric96 replied:
fixed

galak commented:
@ric96 any plans to get back to this?

erwango commented on boards/arm/96b_aerocore2/Kconfig.defconfig:
```diff
@@ -0,0 +1,47 @@
+
+if SPI
+
+config SPI_1
+	default y
+
+config SPI_2
+	default y
+
+config SPI_3
+	default y
+
+config SPI_4
+	default y
```
Please remove these, following recent changes on DT vs Kconfig usage

erwango commented on boards/arm/96b_aerocore2/Kconfig.defconfig:
```diff
@@ -0,0 +1,47 @@
+
+endif #SPI
+
+config I2C_2
+	default y
+	depends on I2C
+
+config ADC_1
+	default y
+	depends on ADC
```
Please remove these, following recent changes on DT vs Kconfig usage

erwango commented on boards/arm/96b_aerocore2/pinmux.c:
```diff
@@ -0,0 +1,111 @@
+	{STM32_PIN_PA2, STM32F4_PINMUX_FUNC_PA2_PWM5_CH3},
+	{STM32_PIN_PA3, STM32F4_PINMUX_FUNC_PA3_PWM5_CH4},
+#endif /* CONFIG_PWM_STM32_5 */
+#ifdef CONFIG_SPI_1
```
#ifdef DT_HAS_NODE(DT_NODELABEL(spi1))

erwango commented on boards/arm/96b_aerocore2/pinmux.c:
```diff
@@ -0,0 +1,111 @@
+	{STM32_PIN_PA6, STM32F4_PINMUX_FUNC_PA6_SPI1_MISO},
+	{STM32_PIN_PA7, STM32F4_PINMUX_FUNC_PA7_SPI1_MOSI},
+#endif	/* CONFIG_SPI_1 */
+#ifdef CONFIG_SPI_2
```
#ifdef DT_HAS_NODE(DT_NODELABEL(spi2))

erwango commented on boards/arm/96b_aerocore2/pinmux.c:
```diff
@@ -0,0 +1,111 @@
+	{STM32_PIN_PB14, STM32F4_PINMUX_FUNC_PB14_SPI2_MISO},
+	{STM32_PIN_PB15, STM32F4_PINMUX_FUNC_PB15_SPI2_MOSI},
+#endif	/* CONFIG_SPI_2 */
+#ifdef CONFIG_SPI_3
```
#ifdef DT_HAS_NODE(DT_NODELABEL(spi3))

erwango commented on boards/arm/96b_aerocore2/pinmux.c:
```diff
@@ -0,0 +1,111 @@
+	{STM32_PIN_PC11, STM32F4_PINMUX_FUNC_PC11_SPI3_MISO},
+	{STM32_PIN_PC12, STM32F4_PINMUX_FUNC_PC12_SPI3_MOSI},
+#endif	/* CONFIG_SPI_3 */
+#ifdef CONFIG_SPI_4
```
#ifdef DT_HAS_NODE(DT_NODELABEL(spi4))

erwango commented on boards/arm/96b_aerocore2/pinmux.c:
```diff
@@ -0,0 +1,111 @@
+	{STM32_PIN_PE13, STM32F4_PINMUX_FUNC_PE13_SPI4_MISO},
+	{STM32_PIN_PE14, STM32F4_PINMUX_FUNC_PE14_SPI4_MOSI},
+#endif	/* CONFIG_SPI_4 */
+#ifdef CONFIG_I2C_2
```
#ifdef DT_HAS_NODE(DT_NODELABEL(i2c2))

erwango commented on boards/arm/96b_aerocore2/pinmux.c:
```diff
@@ -0,0 +1,111 @@
+	{STM32_PIN_PB10, STM32F4_PINMUX_FUNC_PB10_I2C2_SCL},
+	{STM32_PIN_PB11, STM32F4_PINMUX_FUNC_PB11_I2C2_SDA},
+#endif /* CONFIG_I2C_2 */
+#ifdef CONFIG_ADC_1
```
#ifdef DT_HAS_NODE(DT_NODELABEL(adc1))

erwango reviewed (changes requested):
Few more changes requested regarding recent changes in main branch

erwango commented:
Also please fix errors reported heer: https://github.com/zephyrproject-rtos/zephyr/pull/22095#issuecomment-577197493

erwango commented:
@ric96 just miss a few changes before acceptance. FYI: end of merge window is on May-8

erwango reviewed (approved):
Approved, but potential conflict with https://github.com/zephyrproject-rtos/zephyr/pull/24860

erwango commented:
@ric96, can you rebase to fix sanitycheck and also adapt PWM to https://github.com/zephyrproject-rtos/zephyr/pull/24860 ?
(And then I hope we should be good!)

ric96 commented:
now adapted to #24860

galak commented:
> now adapted to #24860

24860 is merged, so you can rebase on master.

erwango commented:
@ric96, I think we can say you're out of luck since master was broken on the commit you rebased.
This is fine now, if you mind rebasing. Thanks for bearing with us.

galak commented:
I made updates to get this to pass CI.  @erwango you good with this?
